File: java/com/google/copybara/git/GitRevision.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git;

import static com.google.common.collect.ImmutableList.toImmutableList;
import static com.google.copybara.git.GitRepository.COPYBARA_FETCH_NAMESPACE;
import static com.google.copybara.git.GitRepository.FULL_REF_NAMESPACE;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.MoreObjects;
import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableListMultimap;
import com.google.common.collect.Iterables;
import com.google.common.collect.ListMultimap;
import com.google.common.flogger.FluentLogger;
import com.google.copybara.exception.RepoException;
import com.google.copybara.git.GitRepository.GitLogEntry;
import com.google.copybara.revision.Revision;
import com.google.copybara.util.CommandOutput;
import java.time.ZonedDateTime;
import java.util.Map.Entry;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Pattern;
import javax.annotation.Nullable;

/** A Git repository reference */
public final class GitRevision implements Revision {

  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
  public static final Pattern COMPLETE_GIT_HASH_PATTERN =
      Pattern.compile("[a-f0-9]{40}|[a-f0-9]{64}");

  private final GitRepository repository;
  private final String hash;
  @Nullable private final String reference;
  private final ListMultimap<String, String> associatedLabels;
  @Nullable private final String reviewReference;
  @Nullable private final String url;
  private String describe;
  private String describeAbbrev;
  private String revisionNumber;
  private Optional<String> fullReference = Optional.empty();

  /**
   * Create a git revision from a complete (40 or 64 characters) git hash string.
   *
   * @param repository git repository that should contain the {@code hash}
   * @param hash the commit hash
   */
  GitRevision(GitRepository repository, String hash) {
    this(
        repository,
        hash,
        /* reviewReference= */ null,
        /* reference= */ null,
        ImmutableListMultimap.of(),
        /* url= */ null);
  }

  /**
   * Create a git revision from a complete (40 or 64 characters) git hash string.
   *
   * @param repository Git repository that should contain the {@code hash}
   * @param hash The commit hash.
   * @param url The url of the repository that the revision comes from.
   */
  GitRevision(GitRepository repository, String hash, String url) {
    this(repository, hash, null, null, ImmutableListMultimap.of(), url);
  }

  /**
   * Create a git revision from a complete (40 or 64 characters) git hash string.
   *
   * @param repository git repository that should contain the {@code hash}
   * @param hash the commit hash
   * @param reviewReference an arbitrary string that allows to keep track of the revision of the
   *     code review being migrated. The hash is not enough because code reviews are mutable and can
   *     go back and forth in to the same revision:
   *     <ul>
   *       <li>V1: hash: abc
   *       <li>V2: hash: cdb
   *       <li>V3: hash: abc. Goes back to the original hash. But we could have already migrated V2.
   *     </ul>
   *
   * @param reference a stable name that describes where this is coming from. Could be a git
   *     reference like 'master'
   * @param associatedLabels labels associated with this reference
   * @param url if present, the url of the repository that the revision comes from
   */
  @VisibleForTesting
  public GitRevision(
      GitRepository repository,
      String hash,
      @Nullable String reviewReference,
      @Nullable String reference,
      ImmutableListMultimap<String, String> associatedLabels,
      @Nullable String url) {
    this.reviewReference = reviewReference;
    Preconditions.checkArgument(
        COMPLETE_GIT_HASH_PATTERN.matcher(hash).matches(),
        "Reference '%s' is not a full git hash (40 characters SHA-1 or 64 characters SHA-256",
        hash);

    this.repository = Preconditions.checkNotNull(repository);
    this.hash = hash;
    this.reference = reference;
    ArrayListMultimap<String, String> labels = ArrayListMultimap.create();
    labels.putAll(associatedLabels);
    // TODO: hsudhof - Remove GIT_SHA1 and GIT_SHORT_SHA1 labels once all instances are updated.
    if (!associatedLabels.containsKey("GIT_SHA1")) {
      labels.put("GIT_SHA1", hash);
      labels.put("GIT_SHORT_SHA1", hash.substring(0, 7));
      labels.put("GIT_HASH", hash);
      labels.put("GIT_SHORT_HASH", hash.substring(0, 7));
    }
    this.associatedLabels = labels;
    this.url = url;
  }

  @Nullable
  @Override
  public String contextReference() {
    return reference;
  }

  @Override
  public String fixedReference() {
    return hash;
  }

  @Override
  public Optional<String> fullReference() {
    if (fullReference.isPresent()) {
      return fullReference;
    }

    if (reference == null || reference.startsWith("refs/")) {
      fullReference = Optional.ofNullable(reference);
    } else {
      try {
        ImmutableList<String> matchingRefs =
            repository
                .showRef(ImmutableList.of(reference, reference + FULL_REF_NAMESPACE))
                .entrySet()
                .stream()
                .filter(e -> e.getKey().startsWith(COPYBARA_FETCH_NAMESPACE + "/refs/"))
                .filter(e -> e.getValue().getHash().equals(hash))
                .map(Entry::getKey)
                .map(GitRevision::getCleanedFullReference)
                .collect(toImmutableList());

        if (matchingRefs.isEmpty()) {
          logger.atInfo().log("No full reference for ref: %s", reference);
        } else {
          // Git allows having branches and tags with the same name. Prioritize tags over branches.
          fullReference =
              matchingRefs.stream()
                  .filter(e -> e.startsWith("refs/tags/"))
                  .findFirst()
                  .or(() -> matchingRefs.stream().findFirst());
        }
      } catch (RepoException e) {
        logger.atWarning().withCause(e).log(
            "Could not determine full reference for ref: %s. Cause: %s", reference, e.getMessage());
        return Optional.empty();
      }
    }
    return fullReference;
  }

  private static String getCleanedFullReference(String originalRef) {
    String fullRef = originalRef;
    String prefix = "refs/copybara_fetch/";
    if (fullRef.startsWith(prefix)) {
      fullRef = fullRef.substring(prefix.length());
    }
    if (fullRef.endsWith(FULL_REF_NAMESPACE)) {
      fullRef = fullRef.substring(0, fullRef.length() - FULL_REF_NAMESPACE.length());
    }

    return fullRef;
  }

  @Override
  public ZonedDateTime readTimestamp() throws RepoException {
    // TODO(malcon): We should be able to skip this for revisions coming from 'git log'.
    ImmutableList<GitLogEntry> entry = repository.log(hash).withLimit(1).run();
    if (entry.isEmpty()) {
      throw new RepoException(String.format("Cannot find '%s' in the git repository", hash));
    }
    return Iterables.getOnlyElement(entry).authorDate();
  }

  @Override
  public String asString() {
    return hash + (reviewReference == null ? "" : " " + reviewReference);
  }

  public String getHash() {
    return hash;
  }

  @Nullable
  public String getReviewReference() {
    return reviewReference;
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .omitNullValues()
        .add("url", url)
        .add("reference", reference)
        .add("hash", this.hash)
        .add("labels", this.associatedLabels)
        .toString();
  }

  @Override
  @Nullable
  public String getUrl() {
    return url;
  }

  @Override
  public Optional<String> getRevisionType() {
    return Optional.of("Git");
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    GitRevision that = (GitRevision) o;
    return Objects.equals(hash, that.hash);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(hash);
  }

  @Override
  public ImmutableListMultimap<String, String> associatedLabels() {
    return ImmutableListMultimap.copyOf(associatedLabels);
  }

  @Override
  public ImmutableList<String> associatedLabel(String label) {
    // We only return git describe if specifically ask for this label
    if (label.equals(GitRepository.GIT_DESCRIBE_CHANGE_VERSION)) {
      return populateDescribe();
    }
    if (label.equals(GitRepository.GIT_SEQUENTIAL_REVISION_NUMBER)) {
      return populateRevisionNumber();
    }
    if (label.equals(GitRepository.GIT_DESCRIBE_ABBREV)) {
      return populateDescribeAbbrev();
    }
    if (label.equals(GitRepository.GIT_TAG_POINTS_AT)) {
      return populateTagPointsAt();
    }
    return ImmutableList.copyOf(associatedLabels.get(label));
  }

  /**
   * Lazily compute describe. In general we could compute this in git.origin, but sometimes
   * there are thousands of changes to be migrated and we end up calling 'git describe' lot of
   * times. Then in the workflow we generally only need one per migration or one per iteration.
   * This is a waste of time (and a real performance issue). Instead we compute here on demand.
   *
   * <p>Synchronized to make sure we do this only once in case it is being called from parallel
   * transformations.
   */
  private synchronized ImmutableList<String> populateDescribe() {
    if (describe == null) {
      try {
        describe = repository.describe(this, false);
      } catch (RepoException e) {
        logger.atWarning().withCause(e).log(
            "Cannot describe version for %s. Using shortened hash.", hash);
        describe = hash.substring(0, 7);
      }
    }
    return ImmutableList.of(describe);
  }

  private synchronized ImmutableList<String> populateTagPointsAt() {
    if (associatedLabels.containsKey(GitRepository.GIT_TAG_POINTS_AT)) {
      return ImmutableList.copyOf(associatedLabels.get(GitRepository.GIT_TAG_POINTS_AT));
    }

    try {
      ImmutableList<String> tags = repository.tagPointsAt(this);
      associatedLabels.putAll(GitRepository.GIT_TAG_POINTS_AT, tags);
      return tags;
    } catch (RepoException e) {
      logger.atWarning().withCause(e).log("Cannot get 'tag --points-to' output for %s", hash);
    }

    return ImmutableList.of();
  }

  private synchronized ImmutableList<String> populateDescribeAbbrev() {
    if (associatedLabels.containsKey(GitRepository.GIT_DESCRIBE_ABBREV)) {
      return ImmutableList.copyOf(associatedLabels.get(GitRepository.GIT_DESCRIBE_ABBREV));
    }

    if (describeAbbrev == null) {
      try {
        describeAbbrev = repository.describeAbbrev(this);
      } catch (RepoException e) {
        logger.atWarning().withCause(e).log("Cannot get closest tag for %s.", hash);
      }
    }
    if (describeAbbrev != null) {
      associatedLabels.put(GitRepository.GIT_DESCRIBE_ABBREV, describeAbbrev);
    }
    return ImmutableList.of(Strings.nullToEmpty(describeAbbrev));
  }

  /** Lazily compute rev number. */
  private synchronized ImmutableList<String> populateRevisionNumber() {
    if (revisionNumber == null) {
      try {
        CommandOutput cmdout = repository.simpleCommand("rev-list", "--count", hash);
        revisionNumber = cmdout.getStdout().trim();
      } catch (RepoException e) {
        logger.atWarning().withCause(e).log(
            "Cannot get revision number for %s. Using shortened hash", hash);
        revisionNumber = "";
      }
    }
    return ImmutableList.of(revisionNumber);
  }

  GitRevision withUrl(String url) {
    return new GitRevision(
        repository,
        hash,
        reviewReference,
        reference,
        ImmutableListMultimap.copyOf(associatedLabels),
        url);
  }

  GitRevision withContextReference(String tag) {
    return new GitRevision(
        repository,
        hash,
        reviewReference,
        tag,
        ImmutableListMultimap.copyOf(associatedLabels),
        url);
  }

  GitRevision withLabels(ImmutableListMultimap<String, String> labels) {
    return new GitRevision(
        repository,
        hash,
        reviewReference,
        reference,
        Revision.addNewLabels(ImmutableListMultimap.copyOf(associatedLabels), labels),
        url);
  }
}

File: src/Copybara.Core/Git/GitRevision.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Text.RegularExpressions;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Revision;

namespace Copybara.Git;

/// <summary>A Git repository reference. Port of <c>com.google.copybara.git.GitRevision</c>.</summary>
public sealed class GitRevision : IRevision
{
    public static readonly Regex CompleteGitHashPattern =
        new("^(?:[a-f0-9]{40}|[a-f0-9]{64})$", RegexOptions.Compiled);

    private readonly GitRepository _repository;
    private readonly GitHashAlgorithm _hashAlgorithm;
    private readonly string _hash;
    private readonly string? _reference;
    private readonly ImmutableListMultimap<string, string> _associatedLabels;
    private readonly string? _reviewReference;
    private readonly string? _url;
    private string? _describe;
    private string? _describeAbbrev;
    private string? _revisionNumber;
    private string? _fullReferenceCache;
    private bool _fullReferenceComputed;

    private readonly object _lock = new();

    /// <summary>
    /// Create a git revision from a complete (40 or 64 characters) git hash string.
    /// </summary>
    public GitRevision(GitRepository repository, string hash)
        : this(repository, hash, null, null, ImmutableListMultimap<string, string>.Empty, null)
    {
    }

    /// <summary>
    /// Create a git revision from a complete git hash string with a url.
    /// </summary>
    public GitRevision(GitRepository repository, string hash, string? url)
        : this(repository, hash, null, null, ImmutableListMultimap<string, string>.Empty, url)
    {
    }

    /// <summary>
    /// Create a git revision from a complete (40 or 64 characters) git hash string.
    /// </summary>
    /// <param name="repository">git repository that should contain the hash</param>
    /// <param name="hash">the commit hash</param>
    /// <param name="reviewReference">an arbitrary string that allows to keep track of the revision of
    ///     the code review being migrated.</param>
    /// <param name="reference">a stable name that describes where this is coming from. Could be a git
    ///     reference like 'master'.</param>
    /// <param name="associatedLabels">labels associated with this reference</param>
    /// <param name="url">if present, the url of the repository that the revision comes from</param>
    public GitRevision(
        GitRepository repository,
        string hash,
        string? reviewReference,
        string? reference,
        ImmutableListMultimap<string, string> associatedLabels,
        string? url)
    {
        _reviewReference = reviewReference;
        Preconditions.CheckArgument(
            CompleteGitHashPattern.IsMatch(hash),
            "Reference '%s' is not a full git hash (40 characters SHA-1 or 64 characters SHA-256)",
            hash);

        _repository = Preconditions.CheckNotNull(repository);
        _hash = hash;
        _hashAlgorithm = GitHashAlgorithmMethods.From(hash);
        _reference = reference;

        var labels = ImmutableListMultimap<string, string>.CreateBuilder();
        var existing = new HashSet<string>();
        foreach (var key in associatedLabels.Keys)
        {
            existing.Add(key);
            labels.PutAll(key, associatedLabels.Get(key));
        }
        string shortHash = hash.Substring(0, 7);
        // TODO: Remove GIT_SHA1 and GIT_SHORT_SHA1 labels once all instances are updated.
        if (_hashAlgorithm == GitHashAlgorithm.Sha1)
        {
            PutIfMissing(labels, existing, "GIT_SHA1", hash);
            PutIfMissing(labels, existing, "GIT_SHORT_SHA1", shortHash);
        }
        PutIfMissing(labels, existing, "GIT_HASH", hash);
        PutIfMissing(labels, existing, "GIT_SHORT_HASH", shortHash);
        _associatedLabels = labels.Build();
        _url = url;
    }

    private static void PutIfMissing(
        ImmutableListMultimap<string, string>.Builder labels,
        HashSet<string> existing,
        string key,
        string value)
    {
        if (existing.Add(key))
        {
            labels.Put(key, value);
        }
    }

    public string? ContextReference() => _reference;

    public string? FixedReference() => _hash;

    public string? FullReference()
    {
        lock (_lock)
        {
            if (_fullReferenceComputed)
            {
                return _fullReferenceCache;
            }

            if (_reference == null || _reference.StartsWith("refs/", StringComparison.Ordinal))
            {
                _fullReferenceCache = _reference;
                _fullReferenceComputed = true;
                return _fullReferenceCache;
            }

            try
            {
                var matchingRefs = new List<string>();
                var refs = _repository.ShowRef(
                    new[] { _reference, _reference + GitRepository.FullRefNamespace });
                foreach (var e in refs)
                {
                    if (!e.Key.StartsWith(
                            GitRepository.CopybaraFetchNamespace + "/refs/", StringComparison.Ordinal))
                    {
                        continue;
                    }
                    if (e.Value.GetHash() != _hash)
                    {
                        continue;
                    }
                    matchingRefs.Add(GetCleanedFullReference(e.Key));
                }

                if (matchingRefs.Count != 0)
                {
                    // Git allows branches and tags with the same name. Prioritize tags over branches.
                    string? tag = matchingRefs.FirstOrDefault(
                        e => e.StartsWith("refs/tags/", StringComparison.Ordinal));
                    _fullReferenceCache = tag ?? matchingRefs[0];
                }
            }
            catch (RepoException)
            {
                // Could not determine full reference; leave unset for this call.
                return null;
            }

            _fullReferenceComputed = true;
            return _fullReferenceCache;
        }
    }

    private static string GetCleanedFullReference(string originalRef)
    {
        string fullRef = originalRef;
        const string prefix = "refs/copybara_fetch/";
        if (fullRef.StartsWith(prefix, StringComparison.Ordinal))
        {
            fullRef = fullRef.Substring(prefix.Length);
        }
        if (fullRef.EndsWith(GitRepository.FullRefNamespace, StringComparison.Ordinal))
        {
            fullRef = fullRef.Substring(0, fullRef.Length - GitRepository.FullRefNamespace.Length);
        }
        return fullRef;
    }

    public DateTimeOffset? ReadTimestamp()
    {
        var entry = _repository.Log(_hash).WithLimit(1).Run();
        if (entry.Count == 0)
        {
            throw new RepoException($"Cannot find '{_hash}' in the git repository");
        }
        return entry[0].AuthorDate;
    }

    public string AsString() =>
        _hash + (_reviewReference == null ? "" : " " + _reviewReference);

    public string GetHash() => _hash;

    public GitHashAlgorithm GetHashAlgorithm() => _hashAlgorithm;

    public string? GetReviewReference() => _reviewReference;

    public override string ToString()
    {
        var sb = new System.Text.StringBuilder("GitRevision{");
        var parts = new List<string>();
        if (_url != null)
        {
            parts.Add($"url={_url}");
        }
        if (_reference != null)
        {
            parts.Add($"reference={_reference}");
        }
        parts.Add($"hash={_hash}");
        sb.Append(string.Join(", ", parts));
        sb.Append('}');
        return sb.ToString();
    }

    public string? GetUrl() => _url;

    public string? GetRevisionType() => "Git";

    public override bool Equals(object? o)
    {
        if (ReferenceEquals(this, o))
        {
            return true;
        }
        if (o is null || GetType() != o.GetType())
        {
            return false;
        }
        var that = (GitRevision)o;
        return _hash == that._hash;
    }

    public override int GetHashCode() => _hash.GetHashCode();

    public ImmutableListMultimap<string, string> AssociatedLabels() => _associatedLabels;

    public IReadOnlyList<string> AssociatedLabel(string label)
    {
        // We only return git describe if specifically asked for this label
        if (label == GitRepository.GitDescribeChangeVersion)
        {
            return PopulateDescribe();
        }
        if (label == GitRepository.GitSequentialRevisionNumber)
        {
            return PopulateRevisionNumber();
        }
        if (label == GitRepository.GitDescribeAbbrev)
        {
            return PopulateDescribeAbbrev();
        }
        if (label == GitRepository.GitTagPointsAt)
        {
            return PopulateTagPointsAt();
        }
        return _associatedLabels.Get(label);
    }

    /// <summary>Lazily compute describe.</summary>
    private IReadOnlyList<string> PopulateDescribe()
    {
        lock (_lock)
        {
            if (_describe == null)
            {
                try
                {
                    _describe = _repository.Describe(this, false);
                }
                catch (RepoException)
                {
                    _describe = _hash.Substring(0, 7);
                }
            }
            return new[] { _describe! };
        }
    }

    private IReadOnlyList<string> PopulateTagPointsAt()
    {
        lock (_lock)
        {
            if (_associatedLabels.ContainsKey(GitRepository.GitTagPointsAt))
            {
                return _associatedLabels.Get(GitRepository.GitTagPointsAt);
            }
            try
            {
                return _repository.TagPointsAt(this);
            }
            catch (RepoException)
            {
                // Cannot get 'tag --points-at' output.
            }
            return Array.Empty<string>();
        }
    }

    private IReadOnlyList<string> PopulateDescribeAbbrev()
    {
        lock (_lock)
        {
            if (_associatedLabels.ContainsKey(GitRepository.GitDescribeAbbrev))
            {
                return _associatedLabels.Get(GitRepository.GitDescribeAbbrev);
            }
            if (_describeAbbrev == null)
            {
                try
                {
                    _describeAbbrev = _repository.DescribeAbbrev(this);
                }
                catch (RepoException)
                {
                    // Cannot get closest tag.
                }
            }
            return new[] { _describeAbbrev ?? "" };
        }
    }

    private IReadOnlyList<string> PopulateRevisionNumber()
    {
        lock (_lock)
        {
            if (_revisionNumber == null)
            {
                try
                {
                    var cmdout = _repository.SimpleCommand("rev-list", "--count", _hash);
                    _revisionNumber = cmdout.GetStdout().Trim();
                }
                catch (RepoException)
                {
                    _revisionNumber = "";
                }
            }
            return new[] { _revisionNumber! };
        }
    }

    internal GitRevision WithUrl(string url) =>
        new(_repository, _hash, _reviewReference, _reference, _associatedLabels, url);

    internal GitRevision WithContextReference(string tag) =>
        new(_repository, _hash, _reviewReference, tag, _associatedLabels, _url);

    internal GitRevision WithLabels(ImmutableListMultimap<string, string> labels) =>
        new(
            _repository,
            _hash,
            _reviewReference,
            _reference,
            IRevision.AddNewLabels(_associatedLabels, labels),
            _url);
}

/// <summary>Supported Git hash algorithms.</summary>
public enum GitHashAlgorithm
{
    /// <summary>SHA-1 hash algorithm (40 characters).</summary>
    Sha1,

    /// <summary>SHA-256 hash algorithm (64 characters).</summary>
    Sha256,
}

/// <summary>Helpers for <see cref="GitHashAlgorithm"/>.</summary>
public static class GitHashAlgorithmMethods
{
    public static int GetLength(this GitHashAlgorithm algorithm) =>
        algorithm == GitHashAlgorithm.Sha1 ? 40 : 64;

    /// <summary>Returns the algorithm corresponding to the length of the given hash.</summary>
    public static GitHashAlgorithm From(string hash) =>
        hash.Length switch
        {
            40 => GitHashAlgorithm.Sha1,
            64 => GitHashAlgorithm.Sha256,
            _ => throw new ArgumentException(
                $"Invalid hash length {hash.Length}: '{hash}'"),
        };
}
